### PR TITLE
Improve build throughput and enable using the Visual Studio 2019 cmake generator

### PR DESCRIPTION
**Pull request by kile0**
**Description**: Describe your changes.
- Improve build throughput time by disabling optimizations for the testing dynamic initializer that lists all of the tests. Locally I found roughly a 25% improvement on both parallel and sequential builds in a common build scenario (`build.bat --config Release --build_wheel --use_openmp`) without impacting performance of the runtime or a noticable impact on testing time. On my machine I experienced the following speedups:
    sequential (in seconds): 1920~ -> 1430~ = 25% improvement
    parallel (in seconds): 1710~ -> 1240~ = 27% improvement
- Add a build option to use the 2019 Visual Studio CMake generator (2017 was left alone as the default generator). 

**Motivation and Context**
- Why is this change required? What problem does it solve?
   Lower build throughput and enabling the option to upgrade to the latest Windows C++ toolset and take advantage of recent toolset optimizations for devs building from source.

- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: tools/ci_build/build.py [hunk 1]
```diff
@@ -146,6 +146,8 @@ def parse_arguments():
     parser.add_argument("--skip_onnx_tests", action='store_true', help="Explicitly disable all onnx related tests")
     parser.add_argument("--enable_msvc_static_runtime", action='store_true', help="Enable static linking of MSVC runtimes.")
     parser.add_argument("--enable_language_interop_ops", action='store_true', help="Enable operator implemented in language other than cpp")
+    parser.add_argument("--cmake_generator", choices=['Visual Studio 15 2017', 'Visual Studio 16 2019'],
```
Comment (snnn) [inline tools/ci_build/build.py]: Please have a note at somewhere to say this is only for windows.

File: onnxruntime/test/providers/cpu/reduction/reduction_test_cases.inl [hunk 2]
```diff
@@ -1,3 +1,4 @@
+#pragma optimize ("", off)
```
Comment (snnn) [inline onnxruntime/test/providers/cpu/reduction/reduction_test_cases.inl]: Does it can be recognized by gcc?

**Comment by snnn** [comment]
Thanks for your contribution for enabling support for "Visual Studio 2019". Currently, we only announced support for Visual Studio 2017 on Windows  platform.  As you are an expert in this area,  could you help me answer one question: what if onnxruntime is compiled by VS 2019, and loaded in python.exe which is compiled by VS 2015 or 2017?  The python.exe comes with a vcruntime140.dll, which is from VS2015. When python.exe starts, it will load that dll first, then onnxruntime. Even onnxruntime depends on VS2019 runtime, it won't load the new vcruntime140.dll again, because the old vcruntime140.dll is already loaded. But these two dlls are different.  Could this setting work? This question puzzled me a long time. Thanks.

File: tools/ci_build/build.py [hunk 3]
```diff
@@ -803,16 +805,16 @@ def main():
         cmake_extra_args = []
         if(is_windows()):
```
Comment (snnn) [inline tools/ci_build/build.py]: Please add a TODO: need to check if cuda_version >=10.1, if cuda is enabled and VS version >=2019

**Comment by kile0** [comment]
> Thanks for your contribution for enabling support for "Visual Studio 2019". Currently, we only announced support for Visual Studio 2017 on Windows platform. As you are an expert in this area, could you help me answer one question: what if onnxruntime is compiled by VS 2019, and loaded in python.exe which is compiled by VS 2015 or 2017? The python.exe comes with a vcruntime140.dll, which is from VS2015. When python.exe starts, it will load that dll first, then onnxruntime. Even onnxruntime depends on VS2019 runtime, it won't load the new vcruntime140.dll again, because the old vcruntime140.dll is already loaded. But these two dlls are different. Could this setting work? This question puzzled me a long time. Thanks.

This is a good point - thank you for catching it! Yes you'd need a 2019 VS redist on your path (https://aka.ms/vs/16/release/vc_redist.x64.exe - I forgot to update the README), but that should be fine. I've been running MLPerf benchmarks with a 2019-built onnxruntime backend on python 3.7 (which in turn uses a 2017 vcruntime140.dll) and haven't run into any problems locating the proper vcruntime140.dll. Does that and passing test suites enough coverage to verify this won't break anyone?

**Comment by snnn** [comment]
The problem is: even the user has installed 2019 VS redist, he may still get the wrong vcruntime140.dll loaded and in use. And based on all the test we are having now,  we haven't find any issue.

**Comment by kile0** [comment]
> The problem is: even the user has installed 2019 VS redist, he may still get the wrong vcruntime140.dll loaded and in use. And based on all the test we are having now, we haven't find any issue.

I ran through the scenario you were concerned with and so I'm reasonably confident this shouldn't be an issue: on a new vm with only the 2017 redist and build tools installed I installed an onnxruntime built with a 2019 toolchain (build.bat --config Release --build_wheel --use_openmp --cmake_generator="Visual Studio 16 2019") and was able to run both an mlperf inference test (https://github.com/mlperf/inference/tree/master/v0.5/classification_and_detection)  and one of the tutorials (https://github.com/microsoft/onnxruntime/blob/master/docs/python/tutorial.rst)

**Comment by snnn** [comment]
please fix the build breaks

File: tools/ci_build/build.py (hunk 3 above)
Comment (kile0) [reply tools/ci_build/build.py]: Added on line 494

**Comment by kile0** [comment]
@snnn the failure looks unrelated to my change:
  Permission denied to modify the 'C:\Program Files\dotnet\sdk\NuGetFallbackFolder' folder.

Are you familiar with this error?

**Pull request merged**